### PR TITLE
Refine dependencies

### LINKED ISSUE #32: Update package dependencies
Over the past few months a few folks have popped in to the PureScript chat and Discourse to ask about the warnings emitted when installing PureScript via NPM, most recently:

https://discourse.purescript.org/t/warnings-when-installing-the-compiler

We should update this package to provide a better first experience to new PureScripters.

### PR DESCRIPTION
**Pull request by imcotton**
I've made minimal changes to address outdated dependencies, they're atomic commits which easy to review or pick individually.

It's mostly by replacing `request` with `make-fetch-happen`, which I made similar changes in purescript/spago#840.

---
resolves #32

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -18,20 +18,18 @@
...
-    "request": "^2.88.0",
     "rimraf": "^2.6.3",
-    "tar": "^4.4.6",
+    "tar": "^6.1.11",
     "which": "^1.3.1",
     "zen-observable": "^0.8.14"
   },
   "devDependencies": {
-    "tap": "^14.2.3"
+    "tap": "^15.1.6"
```
Comment (imcotton) [inline package.json]: ref: tapjs/node-tap#746

**Review by kritzcreek — APPROVED** [review]
Changes look great to me! Thanks for the nice commits making it easy to review.

The only commit I'm not 100% sure about is the one replacing express, but that's mostly because I don't understand what the old implementation was doing. The new one looks reasonable, and if it works for spago I think it ought to work here as well.

**Comment by JordanMartinez** [comment]
> The only commit I'm not 100% sure about is the one replacing express, but that's mostly because I don't understand what the old implementation was doing.

Which commit is that?

**Comment by kritzcreek** [comment]
> Which commit is that?

https://github.com/purescript/npm-installer/pull/33/commits/b3c3656f2826c08db0e92c847b464a3b7ba60543

**Comment by imcotton** [comment]
That change replaces long deprecated `request` module, which is the cause of warning during the installation. Modern `fetch` spec has slightly different interface, thus the diff coming from.

**Comment by JordanMartinez** [comment]
I looked through the docs of the dependency. Here's a summary of what their readmes say in a linear fashion.

[`node-fetch`](https://www.npmjs.com/package/node-fetch) is a lightweight version of `window.fetch` for Node that makes a few tradeoffs from the spec ([v2](https://github.com/node-fetch/node-fetch/blob/HEAD/docs/v2-LIMITS.md) and [v3](https://github.com/node-fetch/node-fetch/blob/HEAD/docs/v3-LIMITS.md). It uses `node-core` streams.

[`minipass-fetch`](https://www.npmjs.com/package/minipass-fetch) matches the API of `node-fetch` but uses "minipass streams" (see below quote) and supports the full set of TLS Options that may be provided to `https.request()` when making `https` requests.

> Minipass streams are faster and more deterministic in their timing contract than node-core streams, making them a better fit for many server-side use cases.

[`minipass`](https://www.npmjs.com/package/minipass) is an implementation of [PassThrough Streams](https://nodejs.org/api/stream.html#stream_class_stream_passthrough).

However `minipass-fetch` doesn't implement some features. Thus, [`make-fetch-happen`](https://www.npmjs.com/package/make-fetch-happen), which is the dependency in question being added here, adds such features:

> - Builds around [minipass-fetch](https://npm.im/minipass-fetch) for the core [fetch API](https://fetch.spec.whatwg.org/) implementation
> - Request pooling out of the box
> - Quite fast, really
> - Automatic HTTP-semantics-aware request retries
> - Cache-fallback automatic "offline mode"
> - Proxy support (http, https, socks, socks4, socks5)
> - Built-in request caching following full HTTP caching rules (Cache-Control, ETag, 304s, cache fallback on error, etc).
> - Customize cache storage with any [Cache API](https://developer.mozilla.org/en-US/docs/Web/API/Cache)-compliant Cache instance. Cache to Redis!
> - Node.js Stream support
> - Transparent gzip and deflate support
> - [Subresource Integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) support

None of the above seems to be a concern to me. 

> ...that's mostly because I don't understand what the old implementation was doing

AFAICT, an HTTP streaming request was being piped into another stream that unpacks the `.tar` file into an unzipped file on the file system.

**Review by JordanMartinez — APPROVED** [review]

**Comment by imcotton** [comment]
Thanks @JordanMartinez for the detailed write up, if only I could adding here:

The major features that missing from native `http` / `https` (for the sake of adoption in here)

- redirection (`curl -L`)
- proxy (`curl -x`)
- body parsing

[request](https://www.npmjs.com/package/request) was a decent 3rd-party module, but unfortunately it has been inactive since **2018** and been put into deprecation in **2020**.

[node-fetch](https://www.npmjs.com/package/node-fetch) is one viable alternative but in **v3** it went to **ESM** only which affects the adoption, currently it focus on its **v4** development.

[make-fetch-happen](https://www.npmjs.com/package/make-fetch-happen) is a re-implementation of `node-fetch`, it maintained by npm, Inc. (now GitHub?), and also facilitates `npm install` which gives strong confidence in this replacement.

**Comment by jamescallumyoung** [comment]
A fresh install today of purescript-installer still throws these warnings. Is it possible to get this PR merged to fix them?

Warnings:
```
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
```

I'm running purescript-installer@v0.2.0 via purescript@v0.14.0 since that's the version required for the exercises in the [purescript-book](https://book.purescript.org).

I'm happy to help fix the current merge conflicts if @imcotton is no longer able to. Just let me know!

**Comment by imcotton** [comment]
Haven't noticed the conflicts afterward, rebased, please let me know if there is anything left.

**Comment by imcotton** [comment]
Changed TAP to ignore coverage ratio during test on CI, @JordanMartinez please help to trigger the workflows, thanks.

**Comment by JordanMartinez** [comment]
This has been published as `v0.3.1`.

**Comment by JordanMartinez** [comment]
Do you mind submitting a PR to the PureScript repo, bumping this dependency's version to `0.3.1`?

**Comment by imcotton** [comment]
I'd happy to, one second : )

**Comment by rhendric** [comment]
Hey @imcotton, I suspect something about this PR is causing #37, but I haven't gone and learned enough about the libraries used to make a proper diagnosis. Could you take a look and see if something jumps out at you?

**Pull request merged**